src/httpaf_lwt_client.ml

```diff
@@ -114,20 +139,108 @@ let single_request resolver ?config ?authenticator ~meth ~headers ?body uri =
     in
     Lwt.wakeup_later notify_finished err
   in
-  let request_body =
-    Httpaf_lwt_unix.Client.request ?config fd req ~error_handler ~response_handler
-  in
+  let request_body, connection =
+    Httpaf.Client_connection.request ?config req ~error_handler ~response_handler in
+  Httpaf_lwt_unix.Client_HTTP_1_1.request ?read_buffer_size:(read_buffer_size config) fd connection ;
   (match body with
    | Some body -> Httpaf.Body.write_string request_body body
    | None -> ());
   Httpaf.Body.close_writer request_body;
   finished
 
+let read_buffer_size = function
+  | Some config -> Some config.H2.Config.read_buffer_size
+  | None -> None
+
+let single_h2_request ?config fd scheme user_pass host meth path headers body =
+  let blen = match body with None -> None | Some x -> Some (String.length x) in
+  let headers = prep_h2_headers headers host user_pass blen in
+  let req = H2.Request.create ~scheme ~headers meth path in
+  Logs.debug (fun m -> m "Sending @[<v 0>%a@]" H2.Request.pp_hum req);
+  let finished, notify_finished = Lwt.wait () in
+  let on_eof response data () =
+    Lwt.wakeup_later notify_finished (Ok (H2 response, data))
+  in
+  let response_handler response response_body =
+    let rec on_read on_eof data bs ~off ~len =
+      let data = data ^ Bigstringaf.substring ~off ~len bs in
+      H2.Body.schedule_read response_body
+        ~on_read:(on_read on_eof data)
+        ~on_eof:(on_eof response (Some data))
+    in
+    H2.Body.schedule_read response_body
+      ~on_read:(on_read on_eof "")
+      ~on_eof:(on_eof response None)
+  in
+  let error_handler e =
+    let err = match e with
+      | `Malformed_response x -> Error (`Msg ("malformed response: " ^ x))
+      | `Invalid_response_body_length _ -> Error (`Msg "invliad response body length")
+      | `Protocol_error (err, msg) -> error_msgf "%a: %s" H2.Error_code.pp_hum err msg
+      | `Exn e -> Error (`Msg ("exception here: " ^ Printexc.to_string e))
+    in
+    Lwt.wakeup_later notify_finished err
+  in
+  let connection =
+    H2.Client_connection.create ?config ?push_handler:None ~error_handler in
+  let request_body =
+    H2.Client_connection.request connection req ~error_handler ~response_handler in
+  Httpaf_lwt_unix.Client_H2.request ?read_buffer_size:(read_buffer_size config) fd connection ;
+  (match body with
+   | Some body -> H2.Body.write_string request_body body
+   | None -> ());
+  H2.Body.close_writer request_body;
+  finished
+
+let alpn_protocol = function
+  | `Plain _ -> None
+  | `Tls tls -> match Tls_lwt.Unix.epoch tls with
+    | Ok { Tls.Core.alpn_protocol= Some "h2"; _ } -> Some `H2
+    | Ok { Tls.Core.alpn_protocol= Some "http/1.1"; _ } -> Some `HTTP_1_1
+    | Ok { Tls.Core.alpn_protocol= None; _ } -> None
+    | Ok { Tls.Core.alpn_protocol= Some _; _ } -> assert false (* XXX(dinosaure): safe? *)
```

Comment:
that's not safe, no. a server may say anything in alpn, so we should play safe -- maybe best to log a warning and return `None`?

what about http2 via TCP (no TLS)? maybe we should follow https://datatracker.ietf.org/doc/html/rfc7540#section-3.2 (and add the desired http 1.1 headers) -- but otoh this would add complexity for no real benefit (since we're doing a single http request anyways)?

Reply:
> what about http2 via TCP (no TLS)? maybe we should follow https://datatracker.ietf.org/doc/html/rfc7540#section-3.2 (and add the desired http 1.1 headers) -- but otoh this would add complexity for no real benefit (since we're doing a single http request anyways)?

This is the purpose of `h2c` but not sure that we really want to handle it.